// @elastic/eui/test-env/components/nav_drawer/nav_drawer_flyout.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EuiNavDrawerFlyout = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _tabbable = _interopRequireDefault(require("tabbable"));

var _services = require("../../services");

var _title = require("../title");

var _nav_drawer_group = require("./nav_drawer_group");

var _list_group = require("../list_group/list_group");

var _focus_trap = require("../focus_trap");

var EuiNavDrawerFlyout = function EuiNavDrawerFlyout(_ref) {
  var className = _ref.className,
      title = _ref.title,
      isCollapsed = _ref.isCollapsed,
      listItems = _ref.listItems,
      wrapText = _ref.wrapText,
      onClose = _ref.onClose,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["className", "title", "isCollapsed", "listItems", "wrapText", "onClose"]);

  var _useState = (0, _react.useState)(),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      menuEl = _useState2[0],
      setMenuEl = _useState2[1];

  var _useState3 = (0, _react.useState)(),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      tabbables = _useState4[0],
      setTabbables = _useState4[1];

  var LABEL = 'navDrawerFlyoutTitle';
  var classes = (0, _classnames.default)('euiNavDrawerFlyout', {
    'euiNavDrawerFlyout-isCollapsed': isCollapsed,
    'euiNavDrawerFlyout-isExpanded': !isCollapsed
  }, className);

  var handleKeyDown = function handleKeyDown(e) {
    if (e.keyCode === _services.keyCodes.ESCAPE) {
      handleClose();
    } else if (e.keyCode === _services.keyCodes.TAB) {
      var tabs = tabbables;

      if (!tabs) {
        tabs = (0, _tabbable.default)(menuEl).filter(function (el) {
          return el.tagName !== 'DIV';
        });
        setTabbables(tabs);
      }

      if (!e.shiftKey && document.activeElement === tabs[tabs.length - 1] || e.shiftKey && document.activeElement === tabs[0]) {
        handleClose();
      }
    }
  };

  var handleClose = function handleClose() {
    var shouldReturnFocus = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    setTabbables(null);
    onClose(shouldReturnFocus);
  };

  return _react.default.createElement("div", (0, _extends2.default)({
    className: classes,
    "aria-labelledby": LABEL,
    onKeyDown: handleKeyDown,
    ref: setMenuEl
  }, rest), _react.default.createElement(_title.EuiTitle, {
    className: "euiNavDrawerFlyout__title",
    tabIndex: "-1",
    size: "xxs"
  }, _react.default.createElement("div", {
    id: LABEL
  }, title)), listItems ? _react.default.createElement(_focus_trap.EuiFocusTrap, {
    returnFocus: false
  }, _react.default.createElement(_nav_drawer_group.EuiNavDrawerGroup, {
    className: "euiNavDrawerFlyout__listGroup",
    ariaLabelledby: LABEL,
    listItems: listItems,
    wrapText: wrapText,
    onClose: function onClose() {
      return handleClose(false);
    }
  })) : null);
};

exports.EuiNavDrawerFlyout = EuiNavDrawerFlyout;
EuiNavDrawerFlyout.propTypes = {
  className: _propTypes.default.string,
  listItems: _list_group.EuiListGroup.propTypes.listItems,
  wrapText: _list_group.EuiListGroup.propTypes.wrapText,

  /**
   * Display a title atop the flyout
   */
  title: _propTypes.default.string,

  /**
   * Toggle the nav drawer between collapsed and expanded
   */
  isCollapsed: _propTypes.default.bool,

  /**
   * Passthrough function to be called when the flyout is closing
   * See ./nav_drawer.js
   */
  onClose: _propTypes.default.func
};
EuiNavDrawerFlyout.defaultProps = {
  isCollapsed: true
};
EuiNavDrawerFlyout.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiNavDrawerFlyout",
  "props": {
    "isCollapsed": {
      "defaultValue": {
        "value": "true",
        "computed": false
      },
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "Toggle the nav drawer between collapsed and expanded"
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "listItems": {
      "type": {
        "name": "custom",
        "raw": "EuiListGroup.propTypes.listItems"
      },
      "required": false,
      "description": ""
    },
    "wrapText": {
      "type": {
        "name": "custom",
        "raw": "EuiListGroup.propTypes.wrapText"
      },
      "required": false,
      "description": ""
    },
    "title": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "Display a title atop the flyout"
    },
    "onClose": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": "Passthrough function to be called when the flyout is closing\nSee ./nav_drawer.js"
    }
  }
};
