// @elastic/charts/dist/chart_types/partition_chart/layout/types/config_types.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PartitionLayout = Object.freeze({
    sunburst: 'sunburst',
    treemap: 'treemap',
});
var Type = (function () {
    function Type(dflt, reconfigurable, documentation) {
        this.documentation = 'string';
        this.dflt = dflt;
        this.reconfigurable = reconfigurable;
        this.documentation = documentation;
    }
    return Type;
}());
var Numeric = (function (_super) {
    __extends(Numeric, _super);
    function Numeric(_a) {
        var dflt = _a.dflt, min = _a.min, max = _a.max, reconfigurable = _a.reconfigurable, documentation = _a.documentation;
        var _this = _super.call(this, dflt, reconfigurable, documentation) || this;
        _this.type = 'number';
        _this.min = min;
        _this.max = max;
        return _this;
    }
    return Numeric;
}(Type));
exports.Numeric = Numeric;
