// @elastic/charts/dist/chart_types/xy_chart/tooltip/tooltip.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var specs_1 = require("../utils/specs");
var geometry_1 = require("../../../utils/geometry");
var accessor_1 = require("../../../utils/accessor");
var series_1 = require("../utils/series");
exports.Y0_ACCESSOR_POSTFIX = ' - lower';
exports.Y1_ACCESSOR_POSTFIX = ' - upper';
function getSeriesTooltipValues(tooltipValues, defaultValue) {
    var seriesTooltipValues = new Map();
    tooltipValues.forEach(function (_a) {
        var value = _a.value, seriesIdentifier = _a.seriesIdentifier, valueAccessor = _a.valueAccessor;
        var seriesValue = defaultValue ? defaultValue : value;
        var current = seriesTooltipValues.get(seriesIdentifier.key) || {};
        var tooltipValue = __assign({ y0: defaultValue, y1: defaultValue }, current);
        if (valueAccessor != null && (valueAccessor === 'y0' || valueAccessor === 'y1')) {
            tooltipValue[valueAccessor] = seriesValue;
        }
        seriesTooltipValues.set(seriesIdentifier.key, tooltipValue);
    });
    return seriesTooltipValues;
}
exports.getSeriesTooltipValues = getSeriesTooltipValues;
function formatTooltip(_a, spec, isHeader, isHighlighted, hasSingleSeries, axisSpec) {
    var color = _a.color, _b = _a.value, x = _b.x, y = _b.y, accessor = _b.accessor, seriesIdentifier = _a.seriesIdentifier;
    var label = series_1.getSeriesName(seriesIdentifier, hasSingleSeries, true, spec);
    if (specs_1.isBandedSpec(spec.y0Accessors) && (specs_1.isAreaSeriesSpec(spec) || specs_1.isBarSeriesSpec(spec))) {
        var _c = spec.y0AccessorFormat, y0AccessorFormat = _c === void 0 ? exports.Y0_ACCESSOR_POSTFIX : _c, _d = spec.y1AccessorFormat, y1AccessorFormat = _d === void 0 ? exports.Y1_ACCESSOR_POSTFIX : _d;
        var formatter = accessor === geometry_1.BandedAccessorType.Y0 ? y0AccessorFormat : y1AccessorFormat;
        label = accessor_1.getAccessorFormatLabel(formatter, label);
    }
    var isFiltered = spec.filterSeriesInTooltip !== undefined ? spec.filterSeriesInTooltip(seriesIdentifier) : true;
    var isVisible = label === '' ? false : isFiltered;
    var value = isHeader ? x : y;
    var tickFormatOptions = spec.timeZone ? { timeZone: spec.timeZone } : undefined;
    return {
        seriesIdentifier: seriesIdentifier,
        valueAccessor: accessor,
        label: label,
        value: axisSpec ? axisSpec.tickFormat(value, tickFormatOptions) : emptyFormatter(value),
        color: color,
        isHighlighted: isHeader ? false : isHighlighted,
        isVisible: isVisible,
    };
}
exports.formatTooltip = formatTooltip;
function emptyFormatter(value) {
    return value;
}
