// @kbn/dev-utils/target/certs.js
"use strict";
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
exports.CA_CERT_PATH = path_1.resolve(__dirname, '../certs/ca.crt');
exports.ES_KEY_PATH = path_1.resolve(__dirname, '../certs/elasticsearch.key');
exports.ES_CERT_PATH = path_1.resolve(__dirname, '../certs/elasticsearch.crt');
exports.ES_P12_PATH = path_1.resolve(__dirname, '../certs/elasticsearch.p12');
exports.ES_P12_PASSWORD = 'storepass';
exports.ES_EMPTYPASSWORD_P12_PATH = path_1.resolve(__dirname, '../certs/elasticsearch_emptypassword.p12');
exports.ES_NOPASSWORD_P12_PATH = path_1.resolve(__dirname, '../certs/elasticsearch_nopassword.p12');
exports.KBN_KEY_PATH = path_1.resolve(__dirname, '../certs/kibana.key');
exports.KBN_CERT_PATH = path_1.resolve(__dirname, '../certs/kibana.crt');
exports.KBN_P12_PATH = path_1.resolve(__dirname, '../certs/kibana.p12');
exports.KBN_P12_PASSWORD = 'storepass';
