// @elastic/eui/es/components/search_bar/search_bar.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

import PropTypes from "prop-types";
import React, { Component } from 'react';
import { isString } from '../../services/predicate';
import { EuiFlexGroup, EuiFlexItem } from '../flex';
import { EuiSearchBox } from './search_box';
import { EuiSearchFilters } from './search_filters';
import { Query } from './query';
export { Query, AST as Ast } from './query';

var parseQuery = function parseQuery(query, props) {
  var schema = undefined;

  if (props.box && props.box.schema && _typeof(props.box.schema) === 'object') {
    schema = props.box.schema;
  }

  var dateFormat = props.dateFormat;
  var parseOptions = {
    schema: schema,
    dateFormat: dateFormat
  };

  if (!query) {
    return Query.parse('', parseOptions);
  }

  return isString(query) ? Query.parse(query, parseOptions) : query;
};

export var EuiSearchBar =
/*#__PURE__*/
function (_Component) {
  _inherits(EuiSearchBar, _Component);

  function EuiSearchBar(props) {
    var _this;

    _classCallCheck(this, EuiSearchBar);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(EuiSearchBar).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "onSearch", function (queryText) {
      try {
        var query = parseQuery(queryText, _this.props);

        _this.notifyControllingParent({
          query: query,
          queryText: queryText,
          error: null
        });

        _this.setState({
          query: query,
          queryText: queryText,
          error: null
        });
      } catch (e) {
        var error = {
          name: e.name,
          message: e.message
        };

        _this.notifyControllingParent({
          query: null,
          queryText: queryText,
          error: error
        });

        _this.setState({
          queryText: queryText,
          error: error
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onFiltersChange", function (query) {
      _this.notifyControllingParent({
        query: query,
        queryText: query.text,
        error: null
      });

      _this.setState({
        query: query,
        queryText: query.text,
        error: null
      });
    });

    var _query = parseQuery(props.defaultQuery || props.query, props);

    _this.state = {
      query: _query,
      queryText: _query.text,
      error: null
    };
    return _this;
  }

  _createClass(EuiSearchBar, [{
    key: "notifyControllingParent",
    value: function notifyControllingParent(newState) {
      var onChange = this.props.onChange;

      if (!onChange) {
        return;
      }

      var oldState = this.state;
      var query = newState.query,
          queryText = newState.queryText,
          error = newState.error;
      var isQueryDifferent = oldState.queryText !== queryText;
      var oldError = oldState.error ? oldState.error.message : null;
      var newError = error ? error.message : null;
      var isErrorDifferent = oldError !== newError;

      if (isQueryDifferent || isErrorDifferent) {
        if (error == null) {
          onChange({
            query: query,
            queryText: queryText,
            error: error
          });
        } else {
          onChange({
            query: null,
            queryText: queryText,
            error: error
          });
        }
      }
    }
  }, {
    key: "renderTools",
    value: function renderTools(tools) {
      if (!tools) {
        return undefined;
      }

      if (Array.isArray(tools)) {
        return tools.map(function (tool) {
          return React.createElement(EuiFlexItem, {
            grow: false,
            key: tool.key == null ? undefined : tool.key
          }, tool);
        });
      }

      return React.createElement(EuiFlexItem, {
        grow: false
      }, tools);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          query = _this$state.query,
          queryText = _this$state.queryText,
          error = _this$state.error;
      var _this$props = this.props,
          box = _this$props.box,
          filters = _this$props.filters,
          toolsLeft = _this$props.toolsLeft,
          toolsRight = _this$props.toolsRight;
      var toolsLeftEl = this.renderTools(toolsLeft);
      var filtersBar = !filters ? undefined : React.createElement(EuiFlexItem, {
        className: "euiSearchBar__filtersHolder",
        grow: false
      }, React.createElement(EuiSearchFilters, {
        filters: filters,
        query: query,
        onChange: this.onFiltersChange
      }));
      var toolsRightEl = this.renderTools(toolsRight);
      return React.createElement(EuiFlexGroup, {
        gutterSize: "m",
        alignItems: "center",
        wrap: true
      }, toolsLeftEl, React.createElement(EuiFlexItem, {
        className: "euiSearchBar__searchHolder",
        grow: true
      }, React.createElement(EuiSearchBox, _extends({}, box, {
        query: queryText,
        onSearch: this.onSearch,
        isInvalid: error != null,
        title: error ? error.message : undefined
      }))), filtersBar, toolsRightEl);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if (nextProps.query && (!prevState.query || typeof nextProps.query !== 'string' && nextProps.query.text !== prevState.query.text || typeof nextProps.query === 'string' && nextProps.query !== prevState.query.text)) {
        var query = parseQuery(nextProps.query, nextProps);
        return {
          query: query,
          queryText: query.text,
          error: null
        };
      }

      return null;
    }
  }]);

  return EuiSearchBar;
}(Component);

_defineProperty(EuiSearchBar, "Query", Query);

EuiSearchBar.propTypes = {
  onChange: PropTypes.func,

  /**
     The initial query the bar will hold when first mounted
     */
  defaultQuery: PropTypes.oneOfType([PropTypes.any.isRequired, PropTypes.string.isRequired]),

  /**
     If you wish to use the search bar as a controlled component, continuously pass the query
     via this prop
     */
  query: PropTypes.oneOfType([PropTypes.any.isRequired, PropTypes.string.isRequired]),

  /**
     Configures the search box. Set `placeholder` to change the placeholder text in the box and
     `incremental` to support incremental (as you type) search.
     */
  box: PropTypes.shape({
    placeholder: PropTypes.string,
    incremental: PropTypes.bool,
    // Boolean values are not meaningful to this component, but are allowed so that other
    // components can use e.g. a true value to mean "auto-derive a schema". See EuiInMemoryTable.
    // Admittedly, this is a bit of a hack.
    schema: PropTypes.oneOfType([PropTypes.shape({
      strict: PropTypes.bool,
      fields: PropTypes.any,
      flags: PropTypes.arrayOf(PropTypes.string.isRequired)
    }).isRequired, PropTypes.bool.isRequired]),
    className: PropTypes.string,
    "aria-label": PropTypes.string,
    "data-test-subj": PropTypes.string
  }),

  /**
     An array of search filters.
     */
  filters: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.shape({
    type: PropTypes.oneOf(["is"]).isRequired,
    field: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    negatedName: PropTypes.string,
    available: PropTypes.func
  }).isRequired, PropTypes.shape({
    type: PropTypes.oneOf(["field_value_selection"]).isRequired,
    field: PropTypes.string,
    name: PropTypes.string.isRequired,
    options: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.shape({
      field: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.number.isRequired, PropTypes.bool.isRequired, PropTypes.shape({
        type: PropTypes.oneOf(["date"]).isRequired,
        raw: PropTypes.any.isRequired,
        granularity: PropTypes.oneOfType([PropTypes.shape({
          es: PropTypes.oneOf(["d", "w", "M", "y"]).isRequired,
          js: PropTypes.oneOf(["day", "week", "month", "year"]).isRequired,
          isSame: PropTypes.func.isRequired,
          start: PropTypes.func.isRequired,
          startOfNext: PropTypes.func.isRequired,
          iso8601: PropTypes.func.isRequired
        }).isRequired, PropTypes.oneOf([undefined])]).isRequired,
        text: PropTypes.string.isRequired,
        resolve: PropTypes.func.isRequired
      }).isRequired]).isRequired,
      name: PropTypes.string,
      view: PropTypes.node
    }).isRequired).isRequired, PropTypes.func.isRequired]).isRequired,
    filterWith: PropTypes.oneOfType([PropTypes.oneOf(["prefix", "includes"]), PropTypes.func.isRequired]),
    cache: PropTypes.number,
    multiSelect: PropTypes.oneOfType([PropTypes.bool.isRequired, PropTypes.oneOf(["and", "or"])]),
    loadingMessage: PropTypes.string,
    noOptionsMessage: PropTypes.string,
    searchThreshold: PropTypes.number,
    available: PropTypes.func
  }).isRequired, PropTypes.shape({
    type: PropTypes.oneOf(["field_value_toggle"]).isRequired,
    field: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.number.isRequired, PropTypes.bool.isRequired, PropTypes.shape({
      type: PropTypes.oneOf(["date"]).isRequired,
      raw: PropTypes.any.isRequired,
      granularity: PropTypes.oneOfType([PropTypes.shape({
        es: PropTypes.oneOf(["d", "w", "M", "y"]).isRequired,
        js: PropTypes.oneOf(["day", "week", "month", "year"]).isRequired,
        isSame: PropTypes.func.isRequired,
        start: PropTypes.func.isRequired,
        startOfNext: PropTypes.func.isRequired,
        iso8601: PropTypes.func.isRequired
      }).isRequired, PropTypes.oneOf([undefined])]).isRequired,
      text: PropTypes.string.isRequired,
      resolve: PropTypes.func.isRequired
    }).isRequired]).isRequired,
    name: PropTypes.string.isRequired,
    negatedName: PropTypes.string,
    available: PropTypes.func,
    operator: PropTypes.oneOf(["eq", "exact", "gt", "gte", "lt", "lte"])
  }).isRequired, PropTypes.shape({
    type: PropTypes.oneOf(["field_value_toggle_group"]).isRequired,
    field: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string.isRequired, PropTypes.number.isRequired, PropTypes.bool.isRequired]).isRequired,
      name: PropTypes.string.isRequired,
      negatedName: PropTypes.string,
      operator: PropTypes.oneOf(["eq", "exact", "gt", "gte", "lt", "lte"])
    }).isRequired).isRequired,
    available: PropTypes.func
  }).isRequired]).isRequired),

  /**
     * Tools which go to the left of the search bar.
     */
  toolsLeft: PropTypes.oneOfType([PropTypes.element.isRequired, PropTypes.arrayOf(PropTypes.element.isRequired).isRequired]),

  /**
     * Tools which go to the right of the search bar.
     */
  toolsRight: PropTypes.oneOfType([PropTypes.element.isRequired, PropTypes.arrayOf(PropTypes.element.isRequired).isRequired]),

  /**
     * Date formatter to use when parsing date values
     */
  dateFormat: PropTypes.any,
  className: PropTypes.string,
  "aria-label": PropTypes.string,
  "data-test-subj": PropTypes.string
};
EuiSearchBar.__docgenInfo = {
  "description": "",
  "methods": [{
    "name": "notifyControllingParent",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "newState",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onSearch",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "queryText",
      "type": null
    }],
    "returns": null
  }, {
    "name": "onFiltersChange",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "query",
      "type": null
    }],
    "returns": null
  }, {
    "name": "renderTools",
    "docblock": null,
    "modifiers": [],
    "params": [{
      "name": "tools",
      "type": null
    }],
    "returns": null
  }],
  "displayName": "EuiSearchBar",
  "props": {
    "onChange": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": ""
    },
    "defaultQuery": {
      "type": {
        "name": "union",
        "value": [{
          "name": "any"
        }, {
          "name": "string"
        }]
      },
      "required": false,
      "description": "The initial query the bar will hold when first mounted"
    },
    "query": {
      "type": {
        "name": "union",
        "value": [{
          "name": "any"
        }, {
          "name": "string"
        }]
      },
      "required": false,
      "description": "If you wish to use the search bar as a controlled component, continuously pass the query\n     via this prop"
    },
    "box": {
      "type": {
        "name": "shape",
        "value": {
          "placeholder": {
            "name": "string",
            "required": false
          },
          "incremental": {
            "name": "bool",
            "required": false
          },
          "schema": {
            "name": "union",
            "value": [{
              "name": "shape",
              "value": {
                "strict": {
                  "name": "bool",
                  "required": false
                },
                "fields": {
                  "name": "any",
                  "required": false
                },
                "flags": {
                  "name": "arrayOf",
                  "value": {
                    "name": "string"
                  },
                  "required": false
                }
              }
            }, {
              "name": "bool"
            }],
            "required": false
          },
          "className": {
            "name": "string",
            "required": false
          },
          "aria-label": {
            "name": "string",
            "required": false
          },
          "data-test-subj": {
            "name": "string",
            "required": false
          }
        }
      },
      "required": false,
      "description": "Configures the search box. Set `placeholder` to change the placeholder text in the box and\n     `incremental` to support incremental (as you type) search."
    },
    "filters": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "union",
          "value": [{
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"is\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": true
              },
              "name": {
                "name": "string",
                "required": true
              },
              "negatedName": {
                "name": "string",
                "required": false
              },
              "available": {
                "name": "func",
                "required": false
              }
            }
          }, {
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"field_value_selection\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": false
              },
              "name": {
                "name": "string",
                "required": true
              },
              "options": {
                "name": "union",
                "value": [{
                  "name": "arrayOf",
                  "value": {
                    "name": "shape",
                    "value": {
                      "field": {
                        "name": "string",
                        "required": false
                      },
                      "value": {
                        "name": "union",
                        "value": [{
                          "name": "string"
                        }, {
                          "name": "number"
                        }, {
                          "name": "bool"
                        }, {
                          "name": "shape",
                          "value": {
                            "type": {
                              "name": "enum",
                              "value": [{
                                "value": "\"date\"",
                                "computed": false
                              }],
                              "required": true
                            },
                            "raw": {
                              "name": "any",
                              "required": true
                            },
                            "granularity": {
                              "name": "union",
                              "value": [{
                                "name": "shape",
                                "value": {
                                  "es": {
                                    "name": "enum",
                                    "value": [{
                                      "value": "\"d\"",
                                      "computed": false
                                    }, {
                                      "value": "\"w\"",
                                      "computed": false
                                    }, {
                                      "value": "\"M\"",
                                      "computed": false
                                    }, {
                                      "value": "\"y\"",
                                      "computed": false
                                    }],
                                    "required": true
                                  },
                                  "js": {
                                    "name": "enum",
                                    "value": [{
                                      "value": "\"day\"",
                                      "computed": false
                                    }, {
                                      "value": "\"week\"",
                                      "computed": false
                                    }, {
                                      "value": "\"month\"",
                                      "computed": false
                                    }, {
                                      "value": "\"year\"",
                                      "computed": false
                                    }],
                                    "required": true
                                  },
                                  "isSame": {
                                    "name": "func",
                                    "required": true
                                  },
                                  "start": {
                                    "name": "func",
                                    "required": true
                                  },
                                  "startOfNext": {
                                    "name": "func",
                                    "required": true
                                  },
                                  "iso8601": {
                                    "name": "func",
                                    "required": true
                                  }
                                }
                              }, {
                                "name": "enum",
                                "value": [{
                                  "value": "undefined",
                                  "computed": true
                                }]
                              }],
                              "required": true
                            },
                            "text": {
                              "name": "string",
                              "required": true
                            },
                            "resolve": {
                              "name": "func",
                              "required": true
                            }
                          }
                        }],
                        "required": true
                      },
                      "name": {
                        "name": "string",
                        "required": false
                      },
                      "view": {
                        "name": "node",
                        "required": false
                      }
                    }
                  }
                }, {
                  "name": "func"
                }],
                "required": true
              },
              "filterWith": {
                "name": "union",
                "value": [{
                  "name": "enum",
                  "value": [{
                    "value": "\"prefix\"",
                    "computed": false
                  }, {
                    "value": "\"includes\"",
                    "computed": false
                  }]
                }, {
                  "name": "func"
                }],
                "required": false
              },
              "cache": {
                "name": "number",
                "required": false
              },
              "multiSelect": {
                "name": "union",
                "value": [{
                  "name": "bool"
                }, {
                  "name": "enum",
                  "value": [{
                    "value": "\"and\"",
                    "computed": false
                  }, {
                    "value": "\"or\"",
                    "computed": false
                  }]
                }],
                "required": false
              },
              "loadingMessage": {
                "name": "string",
                "required": false
              },
              "noOptionsMessage": {
                "name": "string",
                "required": false
              },
              "searchThreshold": {
                "name": "number",
                "required": false
              },
              "available": {
                "name": "func",
                "required": false
              }
            }
          }, {
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"field_value_toggle\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": true
              },
              "value": {
                "name": "union",
                "value": [{
                  "name": "string"
                }, {
                  "name": "number"
                }, {
                  "name": "bool"
                }, {
                  "name": "shape",
                  "value": {
                    "type": {
                      "name": "enum",
                      "value": [{
                        "value": "\"date\"",
                        "computed": false
                      }],
                      "required": true
                    },
                    "raw": {
                      "name": "any",
                      "required": true
                    },
                    "granularity": {
                      "name": "union",
                      "value": [{
                        "name": "shape",
                        "value": {
                          "es": {
                            "name": "enum",
                            "value": [{
                              "value": "\"d\"",
                              "computed": false
                            }, {
                              "value": "\"w\"",
                              "computed": false
                            }, {
                              "value": "\"M\"",
                              "computed": false
                            }, {
                              "value": "\"y\"",
                              "computed": false
                            }],
                            "required": true
                          },
                          "js": {
                            "name": "enum",
                            "value": [{
                              "value": "\"day\"",
                              "computed": false
                            }, {
                              "value": "\"week\"",
                              "computed": false
                            }, {
                              "value": "\"month\"",
                              "computed": false
                            }, {
                              "value": "\"year\"",
                              "computed": false
                            }],
                            "required": true
                          },
                          "isSame": {
                            "name": "func",
                            "required": true
                          },
                          "start": {
                            "name": "func",
                            "required": true
                          },
                          "startOfNext": {
                            "name": "func",
                            "required": true
                          },
                          "iso8601": {
                            "name": "func",
                            "required": true
                          }
                        }
                      }, {
                        "name": "enum",
                        "value": [{
                          "value": "undefined",
                          "computed": true
                        }]
                      }],
                      "required": true
                    },
                    "text": {
                      "name": "string",
                      "required": true
                    },
                    "resolve": {
                      "name": "func",
                      "required": true
                    }
                  }
                }],
                "required": true
              },
              "name": {
                "name": "string",
                "required": true
              },
              "negatedName": {
                "name": "string",
                "required": false
              },
              "available": {
                "name": "func",
                "required": false
              },
              "operator": {
                "name": "enum",
                "value": [{
                  "value": "\"eq\"",
                  "computed": false
                }, {
                  "value": "\"exact\"",
                  "computed": false
                }, {
                  "value": "\"gt\"",
                  "computed": false
                }, {
                  "value": "\"gte\"",
                  "computed": false
                }, {
                  "value": "\"lt\"",
                  "computed": false
                }, {
                  "value": "\"lte\"",
                  "computed": false
                }],
                "required": false
              }
            }
          }, {
            "name": "shape",
            "value": {
              "type": {
                "name": "enum",
                "value": [{
                  "value": "\"field_value_toggle_group\"",
                  "computed": false
                }],
                "required": true
              },
              "field": {
                "name": "string",
                "required": true
              },
              "items": {
                "name": "arrayOf",
                "value": {
                  "name": "shape",
                  "value": {
                    "value": {
                      "name": "union",
                      "value": [{
                        "name": "string"
                      }, {
                        "name": "number"
                      }, {
                        "name": "bool"
                      }],
                      "required": true
                    },
                    "name": {
                      "name": "string",
                      "required": true
                    },
                    "negatedName": {
                      "name": "string",
                      "required": false
                    },
                    "operator": {
                      "name": "enum",
                      "value": [{
                        "value": "\"eq\"",
                        "computed": false
                      }, {
                        "value": "\"exact\"",
                        "computed": false
                      }, {
                        "value": "\"gt\"",
                        "computed": false
                      }, {
                        "value": "\"gte\"",
                        "computed": false
                      }, {
                        "value": "\"lt\"",
                        "computed": false
                      }, {
                        "value": "\"lte\"",
                        "computed": false
                      }],
                      "required": false
                    }
                  }
                },
                "required": true
              },
              "available": {
                "name": "func",
                "required": false
              }
            }
          }]
        }
      },
      "required": false,
      "description": "An array of search filters."
    },
    "toolsLeft": {
      "type": {
        "name": "union",
        "value": [{
          "name": "element"
        }, {
          "name": "arrayOf",
          "value": {
            "name": "element"
          }
        }]
      },
      "required": false,
      "description": "Tools which go to the left of the search bar."
    },
    "toolsRight": {
      "type": {
        "name": "union",
        "value": [{
          "name": "element"
        }, {
          "name": "arrayOf",
          "value": {
            "name": "element"
          }
        }]
      },
      "required": false,
      "description": "Tools which go to the right of the search bar."
    },
    "dateFormat": {
      "type": {
        "name": "any"
      },
      "required": false,
      "description": "Date formatter to use when parsing date values"
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "aria-label": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "data-test-subj": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    }
  }
};
