// @kbn/analytics/target/web/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ReportHTTP", {
  enumerable: true,
  get: function get() {
    return _reporter.ReportHTTP;
  }
});
Object.defineProperty(exports, "Reporter", {
  enumerable: true,
  get: function get() {
    return _reporter.Reporter;
  }
});
Object.defineProperty(exports, "ReporterConfig", {
  enumerable: true,
  get: function get() {
    return _reporter.ReporterConfig;
  }
});
Object.defineProperty(exports, "UiStatsMetricType", {
  enumerable: true,
  get: function get() {
    return _metrics.UiStatsMetricType;
  }
});
Object.defineProperty(exports, "METRIC_TYPE", {
  enumerable: true,
  get: function get() {
    return _metrics.METRIC_TYPE;
  }
});
Object.defineProperty(exports, "Report", {
  enumerable: true,
  get: function get() {
    return _report.Report;
  }
});
Object.defineProperty(exports, "ReportManager", {
  enumerable: true,
  get: function get() {
    return _report.ReportManager;
  }
});
Object.defineProperty(exports, "Storage", {
  enumerable: true,
  get: function get() {
    return _storage.Storage;
  }
});

var _reporter = require("./reporter");

var _metrics = require("./metrics");

var _report = require("./report");

var _storage = require("./storage");
