// @elastic/eui/test-env/components/date_picker/super_date_picker/types.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.quickSelectPanelShape = exports.recentlyUsedRangeShape = exports.commonlyUsedRangeShape = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var commonlyUsedRangeShape = _propTypes.default.shape({
  start: _propTypes.default.string.isRequired,
  end: _propTypes.default.string.isRequired,
  label: _propTypes.default.string.isRequired
});

exports.commonlyUsedRangeShape = commonlyUsedRangeShape;

var recentlyUsedRangeShape = _propTypes.default.shape({
  start: _propTypes.default.string.isRequired,
  end: _propTypes.default.string.isRequired
});

exports.recentlyUsedRangeShape = recentlyUsedRangeShape;

var quickSelectPanelShape = _propTypes.default.shape({
  title: _propTypes.default.string.isRequired,
  content: _propTypes.default.node.isRequired
});

exports.quickSelectPanelShape = quickSelectPanelShape;
